File: packages/app/android/src/main/java/io/invertase/firebase/common/UniversalFirebasePreferences.java
package io.invertase.firebase.common;

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import android.content.Context;
import android.content.SharedPreferences;


import io.invertase.firebase.app.ReactNativeFirebaseApp;

public class UniversalFirebasePreferences {
  private static final String PREFERENCES_FILE = "io.invertase.firebase";
  private static UniversalFirebasePreferences sharedInstance = new UniversalFirebasePreferences();
  private SharedPreferences preferences;

  public static UniversalFirebasePreferences getSharedInstance() {
    return sharedInstance;
  }

  public boolean contains(String key) {
    return getPreferences().contains(key);
  }

  // Boolean
  public void setBooleanValue(String key, boolean value) {
    getPreferences().edit().putBoolean(key, value).apply();
  }

  public boolean getBooleanValue(String key, boolean defaultValue) {
    return getPreferences().getBoolean(key, defaultValue);
  }

  // Int
  public void setIntValue(String key, int value) {
    getPreferences().edit().putInt(key, value).apply();
  }

  public int getIntValue(String key, int defaultValue) {
    return getPreferences().getInt(key, defaultValue);
  }

  // Long
  public void setLongValue(String key, long value) {
    getPreferences().edit().putLong(key, value).apply();
  }

  public long getLongValue(String key, long defaultValue) {
    return getPreferences().getLong(key, defaultValue);
  }

  // String
  public void setStringValue(String key, String value) {
    getPreferences().edit().putString(key, value).apply();
  }

  public String getStringValue(String key, String defaultValue) {
    return getPreferences().getString(key, defaultValue);
  }

  public void clearAll() {
    getPreferences().edit().clear().apply();
  }

  public SharedPreferences.Editor remove(String key){
    return getPreferences().edit().remove(key);
  }

  private SharedPreferences getPreferences() {
    if (preferences == null) {
      preferences = ReactNativeFirebaseApp
        .getApplicationContext()
        .getSharedPreferences(PREFERENCES_FILE, Context.MODE_PRIVATE);
    }
    return preferences;
  }
}

File: packages/app/android/src/main/java/io/invertase/firebase/common/UniversalFirebasePreferences.kt
package io.invertase.firebase.common

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import android.content.Context
import android.content.SharedPreferences
import io.invertase.firebase.app.ReactNativeFirebaseApp

open class UniversalFirebasePreferences {
  private var preferences: SharedPreferences? = null

  open fun contains(key: String?): Boolean = getPreferences().contains(key)

  open fun setBooleanValue(
    key: String?,
    value: Boolean,
  ) {
    getPreferences().edit().putBoolean(key, value).apply()
  }

  open fun getBooleanValue(
    key: String?,
    defaultValue: Boolean,
  ): Boolean = getPreferences().getBoolean(key, defaultValue)

  open fun setIntValue(
    key: String?,
    value: Int,
  ) {
    getPreferences().edit().putInt(key, value).apply()
  }

  open fun getIntValue(
    key: String?,
    defaultValue: Int,
  ): Int = getPreferences().getInt(key, defaultValue)

  open fun setLongValue(
    key: String?,
    value: Long,
  ) {
    getPreferences().edit().putLong(key, value).apply()
  }

  open fun getLongValue(
    key: String?,
    defaultValue: Long,
  ): Long = getPreferences().getLong(key, defaultValue)

  open fun setStringValue(
    key: String?,
    value: String?,
  ) {
    getPreferences().edit().putString(key, value).apply()
  }

  open fun getStringValue(
    key: String?,
    defaultValue: String?,
  ): String? = getPreferences().getString(key, defaultValue)

  open fun clearAll() {
    getPreferences().edit().clear().apply()
  }

  // The caller is responsible for calling apply() or commit() on the returned editor.
  open fun remove(key: String?): SharedPreferences.Editor = getPreferences().edit().remove(key)

  private fun getPreferences(): SharedPreferences {
    if (preferences == null) {
      preferences =
        ReactNativeFirebaseApp
          .getApplicationContext()
          .getSharedPreferences(PREFERENCES_FILE, Context.MODE_PRIVATE)
    }
    return preferences!!
  }

  companion object {
    private const val PREFERENCES_FILE = "io.invertase.firebase"
    private val sharedInstance = UniversalFirebasePreferences()

    @JvmStatic fun getSharedInstance(): UniversalFirebasePreferences = sharedInstance
  }
}
